utils/format_doc.ml

```diff
@@ -0,0 +1,476 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*             Florian Angeletti, projet Cambium, Inria Paris             *)
+(*                                                                        *)
+(*   Copyright 2021 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+
+type box_type =
+  | H
+  | V
+  | HV
+  | HoV
+  | B
+
+type stag = Format.stag
+
+type element =
+  | Text of string
+  | With_size of int
+  | Open_box of { kind: box_type ; indent:int }
+  | Close_box
+  | Open_tag of Format.stag
+  | Close_tag
+  | Open_tbox
+  | Tab_break of { width : int; offset : int }
+  | Set_tab
+  | Close_tbox
+  | Simple_break of { spaces : int; indent: int }
+  | Break of { fits : string * int * string as 'a; breaks : 'a }
+  | Flush of { newline:bool }
+  | Newline
+  | If_newline
```

Comment:
I was surprised by `If_newline` which is an odd construct (very imperative). @Octachron tells me that it is not actually used in the compiler so we could get rid of it if we wanted -- same with `With_size` which is also weird. I don't know if we should rather follow the oddities of Format to the letter, or remove the stuff we don't like and don't need.

Reply:
IIRC `If_newline` predates the form of `Break` here, and I wouldn't be surprised if `If_newline` is not really needed.

If you ever want to include static unicode text/symbols in diagnostics, then I expect `With_size` could be useful. If the unicode you want to emit is known ahead of time, the complicated question of computing its width can be worked around by specifying the size. `Format` itself works well in this way.